# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [2]:
def add_two(x):
    return x+2
print(add_two(3))

#Lamda function
lambda x: x+2
(lambda x: x+2)(3)

5


5

#### Saving to a Variable

In [3]:
f = lambda x,y : x+y+2
f(3,5)

10

#### Multiple Inputs

In [5]:
#multiple inputs not stored as a variable
(lambda x,y,z:x*y*z)(3,5,8)

#stored as a variable
x = lambda x,y,z:x*y*z
x(3,5,8)

120

#### Passing a Lambda into a Function

In [7]:
def multiply(f, num):
    return f(num)
multiply(lambda x: x**2,4)

16

#### Returning a Lambda from a Function

In [12]:
def return_fun():
    def multiply(num):
        return num*2
    return multiply

return_fun()(2)

def return_lamb(b,c):
    return lambda x,a: x + a + b + c
return_lamb(1,2)(4,5)
            #b,c x,a


12

#### If Statements within Lambdas

In [14]:
f = lambda num : num*2 if num > 10 else num+2
print(f(10))
print(f(12))

12
24


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [15]:
f = lambda my_variable: my_variable**3
f(2)

8

In [16]:
sorted([2,6,1231,6,1])

[1, 2, 6, 6, 1231]

In [17]:
help(sorted)

Help on built-in function sorted in module builtins:

sorted(iterable, /, *, key=None, reverse=False)
    Return a new list containing all items from the iterable in ascending order.
    
    A custom key function can be supplied to customize the sort order, and the
    reverse flag can be set to request the result in descending order.



In [20]:
#using lambda with sort to sort dictionary by a key from highest to lowest
k = {
    "name":"Kevin",
    "age":36
}
j = {
    "name":"John",
    "age":41
}
d = {
    "name":"Derek",
    "age":32
}
b = {
    "name":"Brian",
    "age":29
}

my_list = [k,j,d,b]
print(my_list)

print(sorted(my_list, key=lambda person_dict: person_dict["age"]))
print(sorted(my_list, key=lambda person_dict: person_dict["age"], reverse = True))

[{'name': 'Kevin', 'age': 36}, {'name': 'John', 'age': 41}, {'name': 'Derek', 'age': 32}, {'name': 'Brian', 'age': 29}]
[{'name': 'Brian', 'age': 29}, {'name': 'Derek', 'age': 32}, {'name': 'Kevin', 'age': 36}, {'name': 'John', 'age': 41}]
[{'name': 'John', 'age': 41}, {'name': 'Kevin', 'age': 36}, {'name': 'Derek', 'age': 32}, {'name': 'Brian', 'age': 29}]


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [24]:
# map(func, iterable)
# help(map)
#typically you will use prederfined function but you can use lambdas as well

def squared (num, num2):
    if num < 10 and num2 < 10:
        return num**2, num2**2
    else:
        return num, num2

numbers = [4,10,3,2,6]
more_nums = [4,11,20,3,15]

squared_nums_map = map(squared, numbers, more_nums)
list(squared_nums_map)

[(16, 16), (10, 11), (3, 20), (4, 9), (6, 15)]

#### Using Lambda's with Map

In [25]:
squared_nums = list(map(lambda x: x**2 if x<10 else x, numbers))
print(squared_nums)

[16, 10, 9, 4, 36]


In [32]:
squared_nums2 = list(map((lambda x1, y1: (x1**2, y1**2) if x1<10 and y1<10 else (x1,y1)),numbers, more_nums))

squared_nums2

[(16, 16), (10, 11), (3, 20), (4, 9), (6, 15)]

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [39]:
nums = [4,2,6,8]
double_nums = list(map(lambda x: (x*2)-1, nums))
print(double_nums)


[7, 3, 11, 15]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [40]:
help(filter)

Help on class filter in module builtins:

class filter(object)
 |  filter(function or None, iterable) --> filter object
 |  
 |  Return an iterator yielding those items of iterable for which function(item)
 |  is true. If function is None, return the items that are true.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



In [42]:
names = ["Bob", "Andy", "Max", "Evan", "Angelica"]

#we only want names that start with A
def a_names(name):
    if name[0].lower()=="a":
        return True
    else:
        return False
    
new_names_list = list(filter(a_names, names))
new_names_list

['Andy', 'Angelica']

#### Using Lambda's with Filter()

In [43]:
new_names_lamb = list(filter(lambda name: True if name[0].lower()=="a" else False, names))
new_names_lamb

['Andy', 'Angelica']

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [44]:
from statistics import mean
help(mean)

Help on function mean in module statistics:

mean(data)
    Return the sample arithmetic mean of data.
    
    >>> mean([1, 2, 3, 4, 4])
    2.8
    
    >>> from fractions import Fraction as F
    >>> mean([F(3, 7), F(1, 21), F(5, 3), F(1, 3)])
    Fraction(13, 21)
    
    >>> from decimal import Decimal as D
    >>> mean([D("0.5"), D("0.75"), D("0.625"), D("0.375")])
    Decimal('0.5625')
    
    If ``data`` is empty, StatisticsError will be raised.



In [50]:
numbers = [2,7,4,2,1.6,9,4.4,4.9]
below_mean_nums = list(filter(lambda x: True if x<=mean(nums) else False, numbers))
print(mean(numbers))
below_mean_nums

4.3625


[2, 4, 2, 1.6, 4.4, 4.9]

## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [52]:
from functools import reduce
#reduce(function, iterable)
#help(reduce)

list_1 = [2,4,6,8,10]

def subtract_nums(num1, num2):
    return num1 - num2
result = reduce(subtract_nums,list_1)
result

-26

#### Using Lambda's with Reduce()

In [53]:
l_3 = [1,10,2,30]

result_lamb = reduce(lambda x,y: x-y, l_3)
print(result_lamb)

-41


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [54]:
list_2 = [2,4,6,8,10]
result_lamb = reduce(lambda x,y: x*y, list_2)
result_lamb

3840

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

In [60]:
def recursive(i):
    print("hi")
    if i==4:
        return
    i += 1
    recursive(i)

recursive(0)

hi
hi
hi
hi
hi


#### Implementing a Base Case

In [61]:
def add_nums(num):
    #set Base case
    if num <=1:
        print("add_nums(1)=1")
        return num
    else:
        print(f'add_nums({nums}) = {num}+add_nums({num - 1})')
        return num + add_nums(num-1)
add_nums(5)

add_nums([4, 2, 6, 8]) = 5+add_nums(4)
add_nums([4, 2, 6, 8]) = 4+add_nums(3)
add_nums([4, 2, 6, 8]) = 3+add_nums(2)
add_nums([4, 2, 6, 8]) = 2+add_nums(1)
add_nums(1)=1


15

#### Writing a Factorial Function

In [62]:
#5! = 5*4*3*2*1

def factorial(num):
    #set base case
    if num <= 1:
        return 1
    else:
        return num * factorial(num - 1)
factorial(5)

120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [66]:
def subtract(num):
    if num <= 1:
        return 1
    else:
        return num - subtract(num - 1)
subtract(5)

3

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

In [67]:
range(0,10)
for i in range(0,10):
    print(i)

0
1
2
3
4
5
6
7
8
9


#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [68]:
def my_range(stop, start, step = 1):
    while start < stop:
        yield start
        start += step
        
for i in my_range(20,2):
    print(i)

2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19


#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [70]:
def my_range(stop, start, step = 1):
    while start > stop:
        yield start**2
        start -= step
        
for num in my_range(0,10):
    print(num)

100
81
64
49
36
25
16
9
4
1


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [72]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

new_places = list(filter(lambda place: True if place.replace(" ", "")  else False, places))
new_places

['Argentina', 'San Diego', 'Boston', 'New York']

### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [65]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

names = sorted(author, key=lambda last_name: last_name.split(" ")[-1].lower())
print(names)


['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [80]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

temp_convert = list(map(lambda y: (y[0], ((9/5)*y[1] + 32)), places))
temp_convert

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]

### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [86]:
def fib(num):
    if num > 1:
        return fib(num - 1) + fib(num - 2)
    return num


print(fib(5))

5


In [87]:
def fib():
    a, b = 0, 1
    while True:            
        yield a          
        a, b = b, a + b
        
for index, fibonacci_number in zip(range(10), fib()):
     print('{i:3}: {f:3}'.format(i=index, f=fibonacci_number))


  0:   0
  1:   1
  2:   1
  3:   2
  4:   3
  5:   5
  6:   8
  7:  13
  8:  21
  9:  34
